### PR TITLE
Fast-path for disabled logging 

### PR DESCRIPTION
**Pull request by benaadams**
`logger.IsEnabled(LogLevel.Information)` is a double check via generic + interface and is already checked for the timestamp; so can be used to short-circuit the `_logger.RequestStarting` call which doesn't inline; from which the `IsEnabled` test can then be removed.

If the `startTimestamp == 0` then logging was disabled and `_logger.RequestFinished` can be skipped. The test was in the function; but the function was non-inlinable.

Added `logger.IsEnabled(LogLevel.Information)` test to `_logger.RequestFinished` as `startTimestamp` could be non-zero if diagnostics was enabled but logging wasn't.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Hosting/Internal/HostingApplication.cs [hunk 1]
```diff
@@ -34,19 +34,26 @@ public Context CreateContext(IFeatureCollection contextFeatures)
         {
             var httpContext = _httpContextFactory.Create(contextFeatures);
             var diagnoticsEnabled = _diagnosticSource.IsEnabled("Microsoft.AspNetCore.Hosting.BeginRequest");
-            var startTimestamp = (diagnoticsEnabled || _logger.IsEnabled(LogLevel.Information)) ? Stopwatch.GetTimestamp() : 0;
+            var loggingEnabled = _logger.IsEnabled(LogLevel.Information);
+
+            var startTimestamp = (diagnoticsEnabled || loggingEnabled) ? Stopwatch.GetTimestamp() : 0;
 
             var scope = _logger.RequestScope(httpContext);
```
Comment (benaadams) [inline src/Microsoft.AspNetCore.Hosting/Internal/HostingApplication.cs]: @davidfowl I assume the scope can't be put in the `if`?

Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/Internal/HostingApplication.cs]: No.

File: src/Microsoft.AspNetCore.Hosting/Internal/HostingLoggerExtensions.cs [hunk 2]
```diff
@@ -23,21 +23,19 @@ public static IDisposable RequestScope(this ILogger logger, HttpContext httpCont
 
         public static void RequestStarting(this ILogger logger, HttpContext httpContext)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/Internal/HostingLoggerExtensions.cs]: One thing that bugs me about this change is the fact that `RequestStarting` and `RequestFinished` pretty much rely on things that happen in the `HostingApplication`. Feels like those 2 methods should be moved there for proximity.

**Comment by benaadams** [comment]
Refactoring the logging variants would look something like the last commit?

**Comment by benaadams** [comment]
I've I move logging after diagnostics in `DisposeContext` its a bit cleaner?

**Comment by Tratcher** [comment]
https://github.com/aspnet/Hosting/pull/944

**Comment by benaadams** [comment]
> #944

Wouldn't be incompatable...

**Comment by davidfowl** [comment]
I don't think #944 is viable.

**Comment by davidfowl** [comment]
@benaadams Before I merge this thing. It needs comments. This code is very very very hard to follow 😄

**Comment by benaadams** [comment]
@davidfowl comments better?

**Comment by davidfowl** [comment]
Yep!
